fix(core): ensure only cluster policy is updated on new ns

carlosrodfern opened the pull request:
Regular policies are mistakenly modified and applied at cluster level over time

**Purpose of PR?**:

Fixes: #1840

**Does this PR introduce a breaking change?** No

**If the changes in this PR are manually verified, list down the scenarios covered:**:
Built and deployed, and verified that the policy is not applied on different namespaces or even the same namespace with no matching labels.


**Additional information for reviewer?** :


**Checklist:**
- [x] Bug fix. Fixes #1840
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

<!--

The PR title message must follow convention:
`<type>(<scope>): <subject>`.

Where: <br />
- `type` is to define what type of PR is this.
  Most common types are:
    - `feat`      - for new features, not a new feature for build script
    - `fix`       - for bug fixes or improvements, not a fix for build script
    - `chore`     - changes not related to production code
    - `docs`      - changes related to documentation
    - `style`     - formatting, missing semi colons, linting fix etc; no significant production code changes
    - `test`      - adding missing tests, refactoring tests; no production code change
    - `refactor`  - refactoring production code, eg. renaming a variable or function name, there should not be any significant production code changes

- `scope` is a single word that best describes where the changes fit.
    - feature(`monitor`,`enforcer`)
    - test(`tests`, `bdd`)
    - chore(`build`)
- `subject` is a single line brief description of the changes made in the pull request.

-->

Prateeknandle commented on KubeArmor/core/kubeUpdate.go:
```diff
@@ -959,48 +959,6 @@ func (dm *KubeArmorDaemon) WatchK8sPods() {
 }
 
-func matchClusterSecurityPolicyRule(policy tp.SecurityPolicy) bool {
-	hasInOperator := false
```
1. we should not remove `matchClusterSecurityPolicyRule()`, reason is mentioned [here](https://github.com/kubearmor/KubeArmor/issues/1780#issuecomment-2288151891).
2. let's add a check :
```
if len(policy.Spec.Selector.Identities) >0 {
		return false
	}
```
this would indicate that the policy is a `KubeArmorSecurityPolicy`, and not the cluster policy.

Prateeknandle replied:
cc : @daemon1024

carlosrodfern replied:
I have updated the commit. Thank you @Prateeknandle!

carlosrodfern committed:
fix(core): ensure only cluster policy is updated on new ns

The `if` condition in `GetSecurityPolicies(..)` returns true if
`matchClusterSecurityPolicyRule(..)` evaluates to `true`. That function doesn't
check whether the passed policy is a cluster policy, and since the
`matchExpressions` is empty for container policies, it ends up adding one
namespace (whatever comes back in the k8s client response first that hasn't been
added yet) to NamespaceList of all existing container policies, it then returns
`true` and the policy is added to the `GetSecurityPolicies(..)` response. Over
time, as `matchClusterSecurityPolicyRule(..)` is called, the list of
`NamespaceList` in each regular policy keeps increasing, causing the container
policy to be applied in namespaces where was not intended.

The `matchClusterSecurityPolicyRule(..)` is corrected to apply only on cluster
policies.

Fixes: #1840

Signed-off-by: Carlos Rodriguez-Fernandez <carlosrodrifernandez@gmail.com>

Prateeknandle reviewed (approved):
Thanks @carlosrodfern for this quick fix.

DelusionalOptimist reviewed (approved):
LGTM. Thanks @carlosrodfern! :raised_hands:
